### ISSUE TITLE
Unclear error message when maxed out of memory

### ISSUE THREAD

**Issue by craigd7** [spring-cloud/spring-cloud-dataflow-server-cloudfoundry#185]
For a while, I was running into problems deploying a stream through PCF.  One of the applications would fail during deployment, while the others would deploy fine.  This is what I would see in the logs for the problematic application:

```
Starting deployment of 'dataflow-abc123'
Successfully computed status [unknown] for dataflow-abc123
Successfully computed status [failed] for dataflow-abc123
Successfully computed status [failed] for dataflow-abc123
Successfully computed status [failed] for dataflow-abc123
Done uploading bits for dataflow-abc123
```

I would then see the following error:

```
Failed to start app dataflow-abc123

java.lang.IllegalStateException: Application dataflow-abc123 failed during staging
at org.cloudfoundry.util.ExceptionUtils.illegalState(ExceptionUtils.java:59) ~[cloudfoundry-util-2.0.0.RELEASE.jar!/:2.0.0.RELEASE]
java.lang.IllegalStateException: Application datafllow-abc123 failed during staging
at reactor.core.publisher.MonoThenApply$MonoThenApplyManager$SecondSubscriber.onNext(MonoThenApply.java:201) ~[reactor-core-3.0.0.RELEASE.jar!/:na]
at org.cloudfoundry.operations.applications.DefaultApplications.waitForStaging(DefaultApplications.java:1384) ~[cloudfoundry-operations-2.0.0.RELEASE.jar!/:2.0.0.RELEASE]
at reactor.core.publisher.Operators$MonoSubscriber.complete(Operators.java:913) ~[reactor-core-3.0.0.RELEASE.jar!/:na]
at reactor.core.publisher.MonoThenApply$MonoThenApplyManager.onNext(MonoThenApply.java:97) ~[reactor-core-3.0.0.RELEASE.jar!/:na]
at org.cloudfoundry.operations.applications.DefaultApplications.lambda$startApplicationAndWait$114(DefaultApplications.java:1214) ~[cloudfoundry-operations-2.0.0.RELEASE.jar!/:2.0.0.RELEASE]
at reactor.core.publisher.FluxMap$MapSubscriber.onNext(FluxMap.java:122) ~[reactor-core-3.0.0.RELEASE.jar!/:na]
at reactor.core.publisher.FluxFilterFuseable$FilterFuseableSubscriber.onNext(FluxFilterFuseable.java:108) ~[reactor-core-3.0.0.RELEASE.jar!/:na]
at reactor.core.publisher.FluxMap$MapSubscriber.onNext(FluxMap.java:122) ~[reactor-core-3.0.0.RELEASE.jar!/:na]
at reactor.core.publisher.FluxResume$ResumeSubscriber.onNext(FluxResume.java:74) ~[reactor-core-3.0.0.RELEASE.jar!/:na]
...
```

After a lot of trial and error, I found out that increasing the memory allocation for the Spring Cloud Dataflow Server would fix this problem.  (We had it at 1 GB, and had to bump it up to 10 GB).

To avoid similar frustration for others, it would be very beneficial to clarify this error message to let the developer know they need to increase the memory allocation.

**Comment by sabbyanandan** [comment]
Thanks for reporting. We were to address the documentation gap via #174 and we will in the upcoming milestone release. We will also review surfacing the error message appropriately.

**Comment by markpollack** [comment]
I don't see how the error above leads to the remedy being that the SCDF Server itself should have its memory increased, also going from 1GB to 10 GB is quite a jump.  Can you elaborate more?  

We have tested that with 1GB specified you end up with the buildpacks jvm memory calculator passing on `-Xms681574K -XX:MetaspaceSize=104857K -Xss349K -Xmx681574K -XX:MaxMetaspaceSize=104857K` to the SCDF Server.  This would result in PCF itself killing the SCDF server with an error message like this

```
2016-09-30T10:56:39.395-04:00 [APP/0] [OUT] Exit status 137
2016-09-30T10:56:39.442-04:00 [CELL/0] [OUT] Exit status 0
```
and the likely result being in some platform issues that don't correctly account for non-heap memory usage.  Moving to 2GB (doubles those JVM settings) and does not display that behavior.  We do have reports of the SCDF Server throwing a java `OutOfMemoryError` exception, but that isn't the case in what you are reporting, and are digging into that now.  Generally speaking one has to give a fairly unituitive large size of memory (container+jvm) for PCF, which can be controlled on a per app basis per deployment or a general setting for all apps that are deployed onto pcf (see the ref docs) that helps the 'flapping' problem, the launched app getting killed with a 137 exiti status over and over again.

Neverthless, we will look to see if there is anything more we can get from the response from the cf-java client/PCF server wrt to 'java.lang.IllegalStateException: Application datafllow-abc123 failed during staging' to help diagnose.

**Comment by craigd7** [comment]
Hey Mark.  No, the error doesn't give any indication that we are having a memory problem.  And I did not see anything elsewhere in the logs that we ran out of memory.  I have not seen an explicit OutOfMemoryError.  However, we would repeatedly try to deploy the same stream (with 5 applications) and we would continue to see this behavior.

The only thing that allowed us to fix this problem was to increase the size of memory for Spring Cloud Dataflow.  We brought it up to 5GB and the stream deployment worked without problems.  This worked for a while, but then we had to bump it up even more to 10GB.  Looking in PCF, the memory usage was hanging out around 7.5GB.  More recently, I noticed that simply restarting the Spring Cloud Dataflow Server brought the memory down to a more manageable size.

In speaking with others, it sounds like this is an issue in a development environment where there are a lot of streams being created/deployed/undeployed.  I've been told that a production environment that has minimal stream deployments would not face the same issues with memory.

**Comment by markpollack** [comment]
We are still trying to track this down and reproduce it, despite trying for quite some time using PWS, deploying/undeploying dozens of times.  We are now setting up a longer test to try and reproduce.   It is very strange that you get 
```java.lang.IllegalStateException: Application dataflow-abc123 failed during staging```

as the error that goes away after increasing the SCDF Server memory.   If there is anyway you can provide us with a means to reproduce what you are observing, we would love to try it out

**Comment by sabbyanandan** [comment]
@markpollack: Ouch, I just realized I closed this by mistake - sorry about that. I noticed an empty PR tagged with this issue in the waffle board and I assumed it was in a wrong state, so closed it.

@craigd7 - The ongoing investigation is tracked via https://github.com/spring-cloud/spring-cloud-dataflow-server-cloudfoundry/issues/107 - please if you can provide further insights, that'd be super helpful.

**Comment by craigd7** [comment]
@markpollack - it sounds like what you are trying is exactly what we've been doing.  I don't think we were doing anything else unique or unusual.  We have been using version 1.0.0.

**Comment by sabbyanandan** [comment]
The core problem was fixed in [`reactor-netty`](https://github.com/reactor/reactor-netty) and [`cf-java-client`](https://github.com/cloudfoundry/cf-java-client/commit/d3e6229b8dda4e67c5102db99ed55784ed6b206a) projects that the SCDF's CF-server relies upon for orchestrating streams/tasks in Cloud Foundry. Corresponding [cf-deployer](https://github.com/spring-cloud/spring-cloud-deployer-cloudfoundry/pull/116) and [cf-server](https://github.com/spring-cloud/spring-cloud-dataflow-server-cloudfoundry/pulls?q=is%3Apr+is%3Aclosed) PRs include the relevant updates. 

This will be released as part of the upcoming 1.1 RC release.
